File: PythonProject/PhysiologicalDataVisualization/data_processing/DataParser.py
#此模块为数据解析模块
import numpy as np
try:
    from DoubleBufferQueue import DoubleBufferQueue_ParseData_eeg, DoubleBufferQueue_ParseData_ecg, DoubleBufferQueue_ParseData_gsr
except ImportError:
    from data_processing.DoubleBufferQueue import DoubleBufferQueue_ParseData_eeg, DoubleBufferQueue_ParseData_ecg, DoubleBufferQueue_ParseData_gsr
class DataParserEEG():
    def __init__(self):
        self.PARSER_CODE_ATTENTION = 4    #注意力
        self.PARSER_CODE_MEDITATION = 5   #冥想
        self.PARSER_CODE_POOR_SIGNAL = 2
        self.PARSER_CODE_CONFIGURATION = 4
        self.PARSER_CODE_RAW = 128        #0x80
        self.PARSER_CODE_DEBUG_ONE = 132
        self.PARSER_CODE_DEBUG_TWO = 133
        self.PARSER_CODE_EEG_POWER = 131  #0x83

        self.RAW_DATA_BYTE_LENGTH = 2
        self.EEG_DEBUG_ONE_BYTE_LENGTH = 5
        self.EEG_DEBUG_TWO_BYTE_LENGTH = 3
        self.PARSER_SYNC_BYTE = 170       #0xAA
        self.PARSER_EXCODE_BYTE = 85      #0x55
        self.MULTI_BYTE_CODE_THRESHOLD = 127
        self.PARSER_STATE_SYNC = 1
        self.PARSER_STATE_SYNC_CHECK = 2
        self.PARSER_STATE_PAYLOAD_LENGTH = 3
        self.PARSER_STATE_PAYLOAD = 4
        self.PARSER_STATE_CHKSUM = 5

        self.payloadLength = 0
        self.payloadBytesReceived = 0
        self.payloadSum = 0
        self.checksum = 0
        self.payload = list()            #存放一包中的有效数据

        self.parserStatus = self.PARSER_STATE_SYNC
        self.double_queue_parse_data_eeg = DoubleBufferQueue_ParseData_eeg() #创建解析后的数据双缓冲实例
        #self.ten_eeg_dict = {'delta':0, 'theta':0, 'lowalpha':0, 'highalpha':0,
        #                    'lowbeta':0, 'highbeta':0, 'lowgamma':0, 'midgamma':0,
        #                    'meditation':0, 'attention':0,}
        self._flag_get_ten_eeg = False
        self._flag_get_eeg_eight = False

    def parseByte(self, data):
        if self.parserStatus == self.PARSER_STATE_SYNC:
            if (data & 0xFF) == self.PARSER_SYNC_BYTE:
                self.parserStatus = self.PARSER_STATE_SYNC_CHECK
            else:
                return -1
        elif self.parserStatus == self.PARSER_STATE_SYNC_CHECK:
            if (data & 0xFF) == self.PARSER_SYNC_BYTE:
                self.parserStatus = self.PARSER_STATE_PAYLOAD_LENGTH
            else:
                self.parserStatus = self.PARSER_STATE_SYNC
                return -1
        elif self.parserStatus == self.PARSER_STATE_PAYLOAD_LENGTH:
            self.payloadLength = (data & 0x0FF)     #获取一包数据的长度
            self.payloadBytesReceived = 0
            self.payload.clear()
            self.payloadSum = 0
            self.parserStatus = self.PARSER_STATE_PAYLOAD
        elif self.parserStatus == self.PARSER_STATE_PAYLOAD:
            self.payload.insert(self.payloadBytesReceived ,data)      #存储一包的有效数据
            self.payloadBytesReceived = self.payloadBytesReceived + 1 #记录一包数据接收的个数
            self.payloadSum = self.payloadSum + (data & 0xFF)
            if self.payloadBytesReceived < self.payloadLength:
                return 0
            self.parserStatus = self.PARSER_STATE_CHKSUM
        elif self.parserStatus == self.PARSER_STATE_CHKSUM:
            self.checksum = (data & 0xFF)
            self.parserStatus = self.PARSER_STATE_SYNC
            if self.checksum == (self.payloadSum ^ 0xFFFFFFFF) & 0xFF:
                self.parsePacketPayload()
            else:
                return -1

    def parsePacketPayload(self):
        i = 0
        extendedCodeLevel = 0
        code = 0
        valueBytesLength = 0
        signal = 0 
        rawWaveData = 0
        ten_eeg_dict = dict()
        while i < self.payloadLength:
            extendedCodeLevel = extendedCodeLevel + 1
            while self.payload[i] == self.PARSER_EXCODE_BYTE:
                i = i + 1
            code = self.payload[i] & 0xFF
            i = i + 1
            if code > self.MULTI_BYTE_CODE_THRESHOLD:
                valueBytesLength = self.payload[i] & 0xFF
                i = i + 1
            else:
                valueBytesLength = 1
            if code == self.PARSER_CODE_RAW:      #脑波原始数据
                if valueBytesLength == self.RAW_DATA_BYTE_LENGTH:
                    highOrderByte = self.payload[i]
                    lowOrderByte = self.payload[(i+1)]
                    rawWaveData = (highOrderByte << 8) | lowOrderByte
                    if rawWaveData > 32768:
                        rawWaveData = rawWaveData - 65536
                    self.double_queue_parse_data_eeg.writer_raw_eeg(rawWaveData)   #这里存放原始脑波数据
                i = i + valueBytesLength
            else:
                if code == self.PARSER_CODE_POOR_SIGNAL:
                    signal = self.payload[i] & 0xFF
                    i = i + valueBytesLength
                elif code == self.PARSER_CODE_EEG_POWER:       #8种脑电波形
                    if signal == 0:
                        self._flag_get_ten_eeg = True
                        highOrderByte = self.payload[i]
                        midOrderByte = self.payload[i + 1]
                        lowOrderByte = self.payload[i + 2]
                        delta = (highOrderByte << 16) | (midOrderByte << 8) | lowOrderByte
                        ten_eeg_dict['delta'] = delta         #这里存放delta波
                        highOrderByte = self.payload[i + 3]
                        midOrderByte = self.payload[i + 4]
                        lowOrderByte = self.payload[i + 5]
                        theta = (highOrderByte << 16) | (midOrderByte << 8) | lowOrderByte
                        ten_eeg_dict['theta'] = theta        #这里存放theta波
                        highOrderByte = self.payload[i + 6]
                        midOrderByte = self.payload[i + 7]
                        lowOrderByte = self.payload[i + 8]
                        lowalpha = (highOrderByte << 16) | (midOrderByte << 8) | lowOrderByte
                        ten_eeg_dict['lowalpha'] = lowalpha  #这里存放lowalpha波
                        highOrderByte = self.payload[i + 9]
                        midOrderByte = self.payload[i + 10]
                        lowOrderByte = self.payload[i + 11]
                        highalpha = (highOrderByte << 16) | (midOrderByte << 8) | lowOrderByte
                        ten_eeg_dict['highalpha'] = highalpha #这里存放highalpha波
                        highOrderByte = self.payload[i + 12]
                        midOrderByte = self.payload[i + 13]
                        lowOrderByte = self.payload[i + 14]
                        lowbeta = (highOrderByte << 16) | (midOrderByte << 8) | lowOrderByte
                        ten_eeg_dict['lowbeta'] = lowbeta    #这里存放lowbeta波
                        highOrderByte = self.payload[i + 15]
                        midOrderByte = self.payload[i + 16]
                        lowOrderByte = self.payload[i + 17]
                        highbeta = (highOrderByte << 16) | (midOrderByte << 8) | lowOrderByte
                        ten_eeg_dict['highbeta'] = highbeta  #这里存放highbeta波
                        highOrderByte = self.payload[i + 18]
                        midOrderByte = self.payload[i + 19]
                        lowOrderByte = self.payload[i + 20]
                        lowgamma = (highOrderByte << 16) | (midOrderByte << 8) | lowOrderByte
                        ten_eeg_dict['lowgamma'] = lowgamma  #这里存放lowgamma波
                        highOrderByte = self.payload[i + 21]
                        midOrderByte = self.payload[i + 22]
                        lowOrderByte = self.payload[i + 23]
                        midgamma = (highOrderByte << 16) | (midOrderByte << 8) | lowOrderByte
                        ten_eeg_dict['midgamma'] = midgamma  #这里存放midgamma波
                    i = i + valueBytesLength
                elif code == self.PARSER_CODE_ATTENTION:
                    if signal == 0:
                        self._flag_get_ten_eeg = True
                        attention = self.payload[i] & 0xFF
                        ten_eeg_dict['attention'] = attention #这里存放注意力
                    i = i + valueBytesLength
                elif code == self.PARSER_CODE_MEDITATION:
                    if signal == 0:
                        self._flag_get_ten_eeg = True
                        meditation = self.payload[i] & 0xFF
                        ten_eeg_dict['meditation'] = meditation #这里存放冥想度
                    i = i + valueBytesLength
        if self._flag_get_ten_eeg:
            self.double_queue_parse_data_eeg.writer_ten_eeg(ten_eeg_dict)  #将10种信号存放在缓存中          
            self._flag_get_ten_eeg = False
            self._flag_get_eeg_eight = True
        self.parserStatus = self.PARSER_STATE_SYNC
# Used for AD8235, if used change "DataParserECG_AD8235" to "DataParserECG"
class DataParserECG():
    def __init__(self):
        self.SPACE = 0x20
        self.rawWaveData = 0
        self.double_queue_parse_data_ecg = DoubleBufferQueue_ParseData_ecg() #创建解析后的数据双缓冲实例
    def parseByte(self, data):
        if (data >= 0x30 and data <= 0x39) or data == self.SPACE or data == 0x0D or data == 0x0A:
            if data >= 0x30 and data <= 0x39:
                self.rawWaveData = (self.rawWaveData * 10) + (data - 48)
            elif data == self.SPACE or data == 0x0D :
                self.double_queue_parse_data_ecg.writer_raw_ecg(self.rawWaveData)  #这里存放原始心电数据
                self.rawWaveData = 0

# Used for BMD101, if used change "DataParserECG_BMD101" to "DataParserECG"
class DataParserECG_BMD101():
    def __init__(self):
        self.PARSER_CODE_HEARTRATE = 3   #心率值
        self.PARSER_CODE_POOR_SIGNAL = 2
        self.PARSER_CODE_CONFIGURATION = 8
        self.PARSER_CODE_RAW = 128        #0x80
        self.PARSER_CODE_DEBUG_ONE = 132
        self.PARSER_CODE_DEBUG_TWO = 133

        self.RAW_DATA_BYTE_LENGTH = 2
        self.EEG_DEBUG_ONE_BYTE_LENGTH = 5
        self.EEG_DEBUG_TWO_BYTE_LENGTH = 3
        self.PARSER_SYNC_BYTE = 170       #0xAA
        self.PARSER_EXCODE_BYTE = 85      #0x55
        self.MULTI_BYTE_CODE_THRESHOLD = 127
        self.PARSER_STATE_SYNC = 1
        self.PARSER_STATE_SYNC_CHECK = 2
        self.PARSER_STATE_PAYLOAD_LENGTH = 3
        self.PARSER_STATE_PAYLOAD = 4
        self.PARSER_STATE_CHKSUM = 5

        self.payloadLength = 0
        self.payloadBytesReceived = 0
        self.payloadSum = 0
        self.checksum = 0
        self.payload = list()            #存放一包中的有效数据

        self.parserStatus = self.PARSER_STATE_SYNC
        self.double_queue_parse_data_ecg = DoubleBufferQueue_ParseData_ecg() #创建解析后的数据双缓冲实例

    def parseByte(self, data):
        if self.parserStatus == self.PARSER_STATE_SYNC:
            if (data & 0xFF) != self.PARSER_SYNC_BYTE:
                return -1
            self.parserStatus = self.PARSER_STATE_SYNC_CHECK
        elif self.parserStatus == self.PARSER_STATE_SYNC_CHECK:
            if (data & 0xFF) == self.PARSER_SYNC_BYTE:
                self.parserStatus = self.PARSER_STATE_PAYLOAD_LENGTH
            else:
                self.parserStatus = self.PARSER_STATE_SYNC
                return -1
        elif self.parserStatus == self.PARSER_STATE_PAYLOAD_LENGTH:
            self.payloadLength = (data & 0x0FF)     #获取一包数据的长度
            self.payloadBytesReceived = 0
            self.payload.clear()
            self.payloadSum = 0
            self.parserStatus = self.PARSER_STATE_PAYLOAD
        elif self.parserStatus == self.PARSER_STATE_PAYLOAD:
            self.payload.insert(self.payloadBytesReceived ,data)      #存储一包的有效数据
            self.payloadBytesReceived = self.payloadBytesReceived + 1 #记录一包数据接收的个数
            self.payloadSum = self.payloadSum + (data & 0xFF)
            if self.payloadBytesReceived < self.payloadLength:
                return -1
            self.parserStatus = self.PARSER_STATE_CHKSUM
        elif self.parserStatus == self.PARSER_STATE_CHKSUM:
            self.checksum = (data & 0xFF)
            self.parserStatus = self.PARSER_STATE_SYNC
            if self.checksum == (self.payloadSum ^ 0xFFFFFFFF) & 0xFF:
                self.parsePacketPayload()
            else:
                return -1

    def parsePacketPayload(self):
        i = 0
        extendedCodeLevel = 0
        code = 0
        valueBytesLength = 0
        signal = 0 
        rawWaveData = 0
        heartRateValue = 0
        while i < self.payloadLength:
            extendedCodeLevel = extendedCodeLevel + 1
            while self.payload[i] == self.PARSER_EXCODE_BYTE:
                i = i + 1
            code = self.payload[i] & 0xFF
            i = i + 1
            if code > self.MULTI_BYTE_CODE_THRESHOLD:
                valueBytesLength = self.payload[i] & 0xFF
                i = i + 1
            else:
                valueBytesLength = 1
            if code == self.PARSER_CODE_RAW:      #心率原始数据
                if valueBytesLength == self.RAW_DATA_BYTE_LENGTH:
                    highOrderByte = self.payload[i]
                    lowOrderByte = self.payload[(i+1)]
                    rawWaveData = (highOrderByte << 8) | lowOrderByte
                    if rawWaveData > 32768:
                        rawWaveData = rawWaveData - 65536
                    self.double_queue_parse_data_ecg.writer_raw_ecg(rawWaveData)  #这里存放原始心电数据
                i = i + valueBytesLength
            else:
                if code == self.PARSER_CODE_POOR_SIGNAL:
                    signal = self.payload[i] & 0xFF
                    i = i + valueBytesLength
                elif code == self.PARSER_CODE_HEARTRATE:
                    if signal == 0:
                        heartRateValue = self.payload[i] & 0xFF
                        self.double_queue_parse_data_ecg.writer_heartValue(heartRateValue) #这里存放心率值
                    i = i + valueBytesLength
                elif code == self.PARSER_CODE_CONFIGURATION:
                    i = i + valueBytesLength
                elif code == self.PARSER_CODE_DEBUG_ONE:
                    if valueBytesLength == self.EEG_DEBUG_ONE_BYTE_LENGTH:
                        i = i + valueBytesLength
                elif code == self.PARSER_CODE_DEBUG_TWO:
                    if valueBytesLength == self.EEG_DEBUG_TWO_BYTE_LENGTH:
                        i = i + valueBytesLength
        self.parserStatus = self.PARSER_STATE_SYNC

class DataParserGSR():
    def __init__(self):
        self.SPACE = 0x20
        self.rawWaveData = 0
        self.count = 0 #存储多个个数字
        self.dot_flag = False
        self.dot_num_behind = 1
        self.array = list()
        self.double_queue_parse_data_gsr = DoubleBufferQueue_ParseData_gsr() #创建解析后的数据双缓冲实例
    def parseByte(self, data):
        if (data >= 0x30 and data <= 0x39) or data == self.SPACE or data == 0x0D or data == 0x0A or data == 0x2E:
            if data >= 0x30 and data <= 0x39:
                if not self.dot_flag:
                    self.rawWaveData = (self.rawWaveData * 10) + (data - 48)
                elif self.dot_flag:
                    self.rawWaveData = self.rawWaveData + ((data - 48) / np.power(10, self.dot_num_behind))
                    self.dot_num_behind += 1
            elif data == self.SPACE :
                self.dot_flag = False
                self.dot_num_behind = 1
                self.count += 1
                self.array.append(self.rawWaveData)
                self.rawWaveData = 0
            elif data == 0x2E:#小数点
                self.dot_flag = True
            elif data == 0x0D:
                self.count += 1
                self.array.append(self.rawWaveData)
                if self.count == 28:
                    for element in self.array[:10]:
                        self.double_queue_parse_data_gsr.writer_raw_gsr(element)
                self.array.clear()
                self.count = 0
                self.rawWaveData = 0
                self.dot_flag = False
                self.dot_num_behind = 1

File: AndroidProject/SleepMonitor/app/src/main/java/com/example/data/DataParser.java
package com.example.data;

import android.util.Log;

public class DataParser {
    private static final int PARSER_CODE_ATTENTION = 0x04;
    private static final int PARSER_CODE_MEDITATION = 0x05;
    private static final int PARSER_CODE_POOR_SIGNAL = 0x02;
    private static final int PARSER_CODE_EEG_RAW = 0x80;
    private static final int PARSER_CODE_EEG_POWER = 0X83;
    private static final int PARSER_CODE_MPU6050 = 0x06;
    private static final int PARSER_CODE_SHT11 = 0x07;
    private static final int PARSER_CODE_MLX90615 = 0x08;

    private static final int PARSER_SYNC_BYTE = 0xAA;

    private static final int PARSER_STATE_SYNC = 1;
    private static final int PARSER_STATE_SYNC_CHECK = 2;
    private static final int PARSER_STATE_PAYLOAD_LENGTH = 3;
    private static final int PARSER_STATE_PAYLOAD = 4;
    private static final int PARSER_STATE_CHKSUM = 5;

    private int parserStatus;
    private int payloadLength;
    private int payloadBytesReceived;
    private int payloadSum;
    private byte[] payload = new byte[512];

    private EEGPowerDataPack eegPowerDataPack = new EEGPowerDataPack();
    private EEGRawDataPack eegRawDataPack = new EEGRawDataPack();
    private AttentionDataPack attentionDataPack = new AttentionDataPack();
    private MeditationDataPack meditationDataPack = new MeditationDataPack();
    private SignalQualityDataPack signalQualityDataPack = new SignalQualityDataPack();
    private OtherSensorDataPack otherSensorDataPack = new OtherSensorDataPack();

    public static final DataParserResult dataParserResult = new DataParserResult();

    public DataParser(){
        parserStatus = PARSER_STATE_SYNC;
    }
    public int parserByte(byte buffer){
        int returnValue = 0;
        switch (parserStatus){
            case 1:
                if((buffer & 0xFF) != PARSER_SYNC_BYTE) break;
                parserStatus = PARSER_STATE_SYNC_CHECK;
                break;
            case 2:
                if((buffer & 0xFF) == PARSER_SYNC_BYTE)
                    parserStatus = PARSER_STATE_PAYLOAD_LENGTH;
                else
                    parserStatus = PARSER_STATE_SYNC;
                break;
            case 3:
                payloadLength = (buffer & 0xFF);
                payloadBytesReceived = 0;
                payloadSum = 0;
                parserStatus = PARSER_STATE_PAYLOAD;
                break;
            case 4:
                payload[payloadBytesReceived++] = buffer;
                payloadSum += (buffer & 0xFF);
                if(payloadBytesReceived < payloadLength) break;
                parserStatus = PARSER_STATE_CHKSUM;
                break;
            case 5:
                int checkSum = (buffer & 0xFF);
                parserStatus = PARSER_STATE_SYNC;
                if(checkSum != ((~payloadSum) & 0xFF)){
                    returnValue = -2;
                }
                else{
                    returnValue = 1;
                    parsePacketPayload();
                }
                break;
        }
        return returnValue;
    }
    private void parsePacketPayload(){
        int i = 0;
        int valueBytesLength;
        int code;

        while (i < payloadLength){
            code = payload[i++] & 0xFF;
            valueBytesLength = payload[i++] & 0xFF;
            if(code == PARSER_CODE_MPU6050){
                otherSensorDataPack.mpu6050_pitch = ((int)payload[i] << 8 | (int)payload[i + 1] & 0xFF) / 100;
                otherSensorDataPack.mpu6050_roll = ((int)payload[i + 2] << 8 | (int)payload[i + 3] & 0xFF) / 100;
                otherSensorDataPack.mpu6050_yaw = ((int)payload[i + 4] << 8 | (int)payload[i + 5] & 0xFF) / 100;
                i += valueBytesLength;
            }
            if(code == PARSER_CODE_SHT11){
                otherSensorDataPack.sht11_temp = ((int)payload[i] << 8 | (int)payload[i + 1] & 0xFF) / 100;
                otherSensorDataPack.sht11_humi = ((int)payload[i + 2] << 8 | (int)payload[i + 3] & 0xFF) / 100;
                i += valueBytesLength;
            }
            if(code == PARSER_CODE_MLX90615){
                otherSensorDataPack.mlx90615_temp = ((int)payload[i] << 8 | (int)payload[i + 1] & 0xFF) / 100;
                i += valueBytesLength;
                if(!dataParserResult.other_sensor_put(otherSensorDataPack)){
                    Log.d("DataParser", "data parser put error!!!!");
                    return;
                }
            }
            if(code == PARSER_CODE_EEG_RAW){
                int count = 0;
                byte highOrderByte;
                byte lowOrderByte;
                while(count < valueBytesLength) {
                    highOrderByte = payload[i+count];
                    count++;
                    lowOrderByte = payload[(i + count)];
                    count++;
                    eegRawDataPack.rawData = getRawWaveValue(highOrderByte, lowOrderByte);
                    if (eegRawDataPack.rawData > 32768) eegRawDataPack.rawData -= 65536;
                    if(!dataParserResult.eeg_raw_put(eegRawDataPack)){
                        Log.d("DataParser", "data parser put error!!!!");
                        return;
                    }
                }
                i += valueBytesLength;
            }
            else{
                switch (code){
                    case PARSER_CODE_EEG_POWER:
                        byte highOrderByte;
                        byte lowOrderByte;
                        byte midOrderByte;
                        int count = 0;
                        while(count < valueBytesLength){
                            highOrderByte = payload[i + count];
                            count++;
                            midOrderByte = payload[i + count];
                            count++;
                            lowOrderByte = payload[i + count];
                            count++;
                            eegPowerDataPack.delta = getEEGPowerValue(highOrderByte, midOrderByte, lowOrderByte);
                            highOrderByte = payload[i + count];
                            count++;
                            midOrderByte = payload[i + count];
                            count++;
                            lowOrderByte = payload[i + count];
                            count++;
                            eegPowerDataPack.theta = getEEGPowerValue(highOrderByte, midOrderByte, lowOrderByte);
                            highOrderByte = payload[i + count];
                            count++;
                            midOrderByte = payload[i + count];
                            count++;
                            lowOrderByte = payload[i + count];
                            count++;
                            eegPowerDataPack.lowalpha = getEEGPowerValue(highOrderByte, midOrderByte, lowOrderByte);
                            highOrderByte = payload[i + count];
                            count++;
                            midOrderByte = payload[i + count];
                            count++;
                            lowOrderByte = payload[i + count];
                            count++;
                            eegPowerDataPack.highalpha = getEEGPowerValue(highOrderByte, midOrderByte, lowOrderByte);
                            highOrderByte = payload[i + count];
                            count++;
                            midOrderByte = payload[i + count];
                            count++;
                            lowOrderByte = payload[i + count];
                            count++;
                            eegPowerDataPack.lowbeta = getEEGPowerValue(highOrderByte, midOrderByte, lowOrderByte);
                            highOrderByte = payload[i + count];
                            count++;
                            midOrderByte = payload[i + count];
                            count++;
                            lowOrderByte = payload[i + count];
                            count++;
                            eegPowerDataPack.highbeta = getEEGPowerValue(highOrderByte, midOrderByte, lowOrderByte);
                            highOrderByte = payload[i + count];
                            count++;
                            midOrderByte = payload[i + count];
                            count++;
                            lowOrderByte = payload[i + count];
                            count++;
                            eegPowerDataPack.lowgamma = getEEGPowerValue(highOrderByte, midOrderByte, lowOrderByte);
                            highOrderByte = payload[i + count];
                            count++;
                            midOrderByte = payload[i + count];
                            count++;
                            lowOrderByte = payload[i + count];
                            count++;
                            eegPowerDataPack.midgamma = getEEGPowerValue(highOrderByte, midOrderByte, lowOrderByte);
                            if(!dataParserResult.eeg_power_put(eegPowerDataPack)){
                                Log.d("DataParser", "data parser put error!!!!");
                                return;
                            }
                        }
                        i += valueBytesLength;
                        break;
                    case PARSER_CODE_ATTENTION:
                        int attention_count = 0;
                        byte attention_orderByte;
                        while(attention_count < valueBytesLength){
                            attention_orderByte = payload[i + attention_count];
                            attention_count++;
                            attentionDataPack.attention = attention_orderByte;
                            if(!dataParserResult.attention_put(attentionDataPack)){
                                Log.d("DataParser", "data parser put error!!!!");
                                return;
                            }
                        }
                        i += valueBytesLength;
                        break;
                    case PARSER_CODE_MEDITATION:
                        int meditation_count = 0;
                        byte meditation_orderByte;
                        while(meditation_count < valueBytesLength){
                            meditation_orderByte = payload[i + meditation_count];
                            meditation_count++;
                            meditationDataPack.meditation = meditation_orderByte;
                            if(!dataParserResult.meditation_put(meditationDataPack)){
                                Log.d("DataParser", "data parser put error!!!!");
                                return;
                            }
                        }
                        i += valueBytesLength;
                        break;
                    case PARSER_CODE_POOR_SIGNAL:
                        int signal_count = 0;
                        byte signal_orderByte;
                        while(signal_count < valueBytesLength){
                            signal_orderByte = payload[i+signal_count];
                            signal_count++;
                            signalQualityDataPack.signal_quality = signal_orderByte;
                            if(!dataParserResult.signal_quality_put(signalQualityDataPack)){
                                Log.d("DataParser", "data parser put error!!!!");
                                return;
                            }
                        }
                        i += valueBytesLength;
                        break;
                    default: break;
                }
            }
        }
        parserStatus = PARSER_STATE_SYNC;
    }
    private int getRawWaveValue(byte highOrderByte, byte lowOrderByte){
        int hi = (int)highOrderByte;
        int lo = ((int)lowOrderByte) & 0xFF;
        return (hi << 8) | lo;
    }
    private int getEEGPowerValue(byte highOrderByte, byte midOrderByte, byte lowOrderByte){
        int hi = (int)highOrderByte;
        int mi = (int)midOrderByte;
        int lo = ((int)lowOrderByte) & 0xFF;
        return (hi << 16) | (mi << 8) | lo;
    }

    public class EEGPowerDataPack{
        public int delta;
        public int theta;
        public int lowalpha;
        public int highalpha;
        public int lowbeta;
        public int highbeta;
        public int lowgamma;
        public int midgamma;
    }
    public class AttentionDataPack{
        public int attention;
    }
    public class MeditationDataPack{ ;
        public int meditation;
    }
    public class SignalQualityDataPack{
        public int signal_quality;
    }
    public class EEGRawDataPack{
        public int rawData;
    }
    public class OtherSensorDataPack{
        public float mpu6050_pitch;
        public float mpu6050_roll;
        public float mpu6050_yaw;
        public float sht11_temp;
        public float sht11_humi;
        public float mlx90615_temp;
    }
}